src/v/storage/segment_index.h

```diff
@@ -88,6 +93,13 @@ class segment_index {
     }
     bool non_data_timestamps() const { return _state.non_data_timestamps; }
 
+    void set_retention_timestamp(model::timestamp t) {
+        _retention_timestamp = t;
+    }
+    model::timestamp retention_timestamp() const {
+        return _retention_timestamp.value_or(_state.max_timestamp);
```

Comment:
Should we also gate this on `storage_ignore_timestamps_in_future_sec`, in case we switch back from ignoring timestamps and no longer want to rely on `_retention_timestamp`?

Reply:
I had mainly been thinking of this as a one shot thing, but you're right that it's straightforward to make it toggleable by adding a check here, so I've gone ahead and done that.